### PR TITLE
Add support for named and explicit indexes

### LINKED ISSUE #171: Add support for pinning a package to a specific index
Discussed this with Armin -- pip doesn't support it, and it seems like a big problem? If you have an internal index, but also want to get some packages from PyPI, there's no way to ensure that your internal packages come from your internal index. Packages on PyPI could even shadow them.

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR adds a first-class API for defining registry indexes, beyond our existing `--index-url` and `--extra-index-url` setup.

Specifically, you now define indexes like so in a `uv.toml` or `pyproject.toml` file:

```toml
[[tool.uv.index]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cu121"
```

You can also provide indexes via `--index` and `UV_INDEX`, and override the default index with `--default-index` and `UV_DEFAULT_INDEX`.

### Index priority

Indexes are prioritized in the order in which they're defined, such that the first-defined index has highest priority.

Indexes are also inherited from parent configuration (e.g., the user-level `uv.toml`), but are placed after any indexes in the current project, matching our semantics for other array-based configuration values.

You can mix `--index` and `--default-index` with the legacy `--index-url` and `--extra-index-url` settings; the latter two are merely treated as unnamed `[[tool.uv.index]]` entries.

### Index pinning

If an index includes a name (which is optional), it can then be referenced via `tool.uv.sources`:

```toml
[[tool.uv.index]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cu121"

[tool.uv.sources]
torch = { index = "pytorch" }
```

If an index is marked as `explicit = true`, it can _only_ be used via such references, and will never be searched implicitly:

```toml
[[tool.uv.index]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cu121"
explicit = true

[tool.uv.sources]
torch = { index = "pytorch" }
```

Indexes defined outside of the current project (e.g., in the user-level `uv.toml`) can _not_ be explicitly selected.

(As of now, we only support using a single index for a given `tool.uv.sources` definition.)

### Default index

By default, we include PyPI as the default index. This remains true even if the user defines a `[[tool.uv.index]]` -- PyPI is still used as a fallback. You can mark an index as `default = true` to (1) disable the use of PyPI, and (2) bump it to the bottom of the prioritized list, such that it's used only if a package does not exist on a prior index:

```toml
[[tool.uv.index]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cu121"
default = true
```

### Name reuse

If a name is reused, the higher-priority index with that name is used, while the lower-priority indexes are ignored entirely. 

For example, given:

```toml
[[tool.uv.index]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cu121"

[[tool.uv.index]]
name = "pytorch"
url = "https://test.pypi.org/simple"
```

The `https://test.pypi.org/simple` index would be ignored entirely, since it's lower-priority than `https://download.pytorch.org/whl/cu121` but shares the same name.

Closes #171.

## Future work

- Users should be able to provide authentication for named indexes via environment variables.
- `uv add` should automatically write `--index` entries to the `pyproject.toml` file.
- Users should be able to provide multiple indexes for a given package, stratified by platform:
```toml
[tool.uv.sources]
torch = [
  { index = "cpu", markers = "sys_platform == 'darwin'" },
  { index = "gpu", markers = "sys_platform != 'darwin'" },
]
```
- Users should be able to specify a proxy URL for a given index, to avoid writing user-specific URLs to a lockfile:
```toml
[[tool.uv.index]]
name = "test"
url = "https://private.org/simple"
proxy = "http://<omitted>/pypi/simple"
```

### CODE REVIEW

**Comment by charliermarsh** [comment]
This still needs docs (beyond the inline documentation), but I'd like to align on the semantics described in the PR summary first.

File: crates/distribution-types/src/index_source.rs
Comment (charliermarsh) [inline crates/distribution-types/src/index_source.rs:57]: I left this in for now but can remove before merging. Eventually I want this to support `--find-links`.

**Comment by zanieb** [comment]
> Any indexes defined via [[tool.uv.index]] take priority over any indexes defined via --index-url or --extra-index-url.
> 
> (This is problematic, since as of now, there's no way to provide a [[tool.uv.index]] on the command line, so you can never override in-file configuration with command-line configuration.)

Yeah this seems wrong. Why is it this way?

Do we need an explicit `explicit` tag or can we just use the presence of a name? I guess the name is useful for providing credentials externally so... I guess the explicit tag is important. 

Do we set the explicit tag during a `uv add --index-url <url> <pkg>` operation?

How does index pinning work for transitive dependencies?

How can we teach the PyPI fallback behavior? Are there any bad user experiences where we could suggest adding `default = true` to their index definition?

**Comment by charliermarsh** [comment]
> Yeah this seems wrong. Why is it this way?

Well... we could just invert the priority, if that's what you mean (such that the new index stuff comes last, and the legacy arguments come first). Then the CLI arguments would work as expected. As-is, though, we don't have a way to differentiate between indexes passed on the command line (with `--index-url` or `--extra-index-url`) and indexes defined in a file via `index-url` and `extra-index-url`... So, like, we can't have `index-url` from the `uv.toml` be lower priority than `tool.uv.index`, but `--index-url` be higher priority. Alternatively, we could add a new command-line argument (`--index`?) for these `tool.uv.index` sources, which would also solve the problem?

> Do we need an explicit explicit tag or can we just use the presence of a name? I guess the name is useful for providing credentials externally so... I guess the explicit tag is important.

Yeah roughly this.

> Do we set the explicit tag during a uv add --index-url <url> <pkg> operation?

We don't as of this PR... We can though. I think we probably should? I guess by default we add the index, and make it explicit for that package?

> How does index pinning work for transitive dependencies?

It doesn't have any effect on transitive dependencies. I don't know that it can or should, honestly, because transitive dependencies can be required from multiple different first-party dependencies that could come from different indexes. I think this would be extremely hard to implement correctly and could lead to confusing behavior.

**Comment by inflation** [comment]
I prefer to make `explicit=true` the default behavior, since the first foot gun in the current situation after adding `extra-index-url`, are packages in the `pytorch` channel taking precedence over `pypi`.

**Comment by smphhh** [comment]
> > How does index pinning work for transitive dependencies?
> 
> It doesn't have any effect on transitive dependencies. I don't know that it can or should, honestly, because transitive dependencies can be required from multiple different first-party dependencies that could come from different indexes. I think this would be extremely hard to implement correctly and could lead to confusing behavior.

Did you consider allowing pinning packages to indexes using glob-patterns as mentioned [here](https://github.com/astral-sh/uv/issues/171#issuecomment-2302093785)? AFAIK that is the only solution for pinning transitive dependencies that is straightforward enough but still useful for a set of real use-cases.

**Comment by charliermarsh** [comment]
> Did you consider allowing pinning packages to indexes using glob-patterns as mentioned https://github.com/astral-sh/uv/issues/171#issuecomment-2302093785? AFAIK that is the only solution for pinning transitive dependencies that is straightforward enough but still useful for a set of real use-cases.

There's nothing stopping us from supporting that in addition to the schema described above. It strikes me as somewhat backwards but I understand why it's useful. My only concern is that we're complicating the schema and creating two ways to assign a package to an index.

**Comment by zanieb** [comment]
I wonder if we should have a separate schema for pinning transitive packages to a defined index? It could allow globs as well. I'm not sure what all the trade-offs are.

**Comment by charliermarsh** [comment]
@zanieb -- That's a good call. Instead of putting this on the index schema, we could have a separate table for assigning packages to named indexes.

**Comment by charliermarsh** [comment]
I'll probably be pushing to this branch a lot over the next few days, so you may want to unsubscribe if the notifications are annoying!

**Comment by j178** [comment]
While playing with this, I found an issue that the lockfile is always ignored due missing remote index:
```console
DEBUG Ignoring existing lockfile due to missing remote index: `idna` `3.4` from `https://download.pytorch.org/whl/cu121`
```

An reproduction:
```console
$ cat pyproject.toml
[project]
name = "foo"
version = "0.1.0"
requires-python = ">=3.12"
dependencies = [
    "idna>=3.4",
]

[[tool.uv.index]]
name = "pypi"
url = "https://pypi.org/simple"
default = true

[[tool.uv.index]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cu121"
explicit = true

[tool.uv.sources]
idna = { index = 'pytorch' }

$ cargo run -- lock
$ cargo run -- lock
```

Maybe this part of code should be updated?
https://github.com/astral-sh/uv/blob/805f1bd6f548f142acd592e5089597b31a012a7c/crates/uv-resolver/src/lock/mod.rs#L1046-L1067

**Comment by charliermarsh** [comment]
Thanks, good catch.

File: crates/uv/tests/show_settings.rs
Comment (konstin) [inline crates/uv/tests/show_settings.rs:3682]: Stale comment

**Review by konstin — APPROVED** [review]

File: crates/distribution-types/src/index.rs
Comment (zanieb) [inline crates/distribution-types/src/index.rs:29]: Perhaps

```suggestion
    /// Explicit indexes will _only_ be used when explicitly requested via a `[tool.uv.sources]`
```

File: crates/distribution-types/src/index.rs
Comment (zanieb) [inline crates/distribution-types/src/index.rs:51]: I thought it was the opposite?

File: crates/distribution-types/src/index.rs
Comment (zanieb) [inline crates/distribution-types/src/index.rs:73]: Is this addressed elsewhere?

File: crates/distribution-types/src/index.rs
Comment (zanieb) [inline crates/distribution-types/src/index.rs:117]: I presume this check is intended to avoid collisions with usernames after the scheme?

File: crates/distribution-types/src/index_url.rs
Comment (zanieb) [inline crates/distribution-types/src/index_url.rs:455]: I don't quite follow the "pip perspective" part.

File: crates/distribution-types/src/index.rs
Comment (charliermarsh) [reply crates/distribution-types/src/index.rs:73]: No it's a TODO. I intentionally left it in but I don't feel strongly.

File: crates/distribution-types/src/index.rs
Comment (charliermarsh) [reply crates/distribution-types/src/index.rs:51]: Yeah sorry, this comment looks stale. It is the opposite.

File: crates/distribution-types/src/index.rs
Comment (charliermarsh) [reply crates/distribution-types/src/index.rs:117]: I think I just wouldn't want to detect `https://foo=bar.com` as a name `https://foo`.

File: crates/pypi-types/src/requirement.rs
Comment (zanieb) [reply crates/pypi-types/src/requirement.rs:322]: The comment seems wrong as-is

File: crates/uv-settings/src/settings.rs
Comment (zanieb) [inline crates/uv-settings/src/settings.rs:772]: Same comment as earlier.

File: crates/uv-settings/src/settings.rs
Comment (zanieb) [inline crates/uv-settings/src/settings.rs:754]: Should we refer to index strategies here too?

File: crates/uv/tests/lock.rs
Comment (zanieb) [inline crates/uv/tests/lock.rs:11825]: Just "named" right? There's no explicit declaration here.

File: crates/uv/tests/lock.rs
Comment (zanieb) [inline crates/uv/tests/lock.rs:11893]: Might be safer to use `example.com` or something so the test fails if it's used?

File: crates/uv/tests/lock.rs
Comment (zanieb) [inline crates/uv/tests/lock.rs:11897]: What if I provided the named index _again_ via the CLI?

File: docs/concepts/dependencies.md
Comment (zanieb) [inline docs/concepts/dependencies.md:73]: I worry the use of "explicit" here will be conflated with the `explicit` index marker

File: docs/configuration/environment.md
Comment (zanieb) [inline docs/configuration/environment.md:6]: Is "base index" right right idea here?

File: docs/configuration/environment.md
Comment (zanieb) [inline docs/configuration/environment.md:8]: Can't be multiple values right? I think these might be roughly swapped but both need some work.

File: docs/configuration/indexes.md
Comment (zanieb) [inline docs/configuration/indexes.md:5]:
```suggestion
private indexes, via the `[[tool.uv.index]]` configuration option (and `--index`, the analogous
```

File: docs/configuration/indexes.md
Comment (zanieb) [inline docs/configuration/indexes.md:15]: Would avoid the use of "explicit" here

File: docs/configuration/indexes.md
Comment (zanieb) [reply docs/configuration/indexes.md:17]: I'd discuss the expected format somewhere else, I think. It seems too irrelevant to most users to put here.

File: docs/configuration/index.md
Comment (zanieb) [inline docs/configuration/index.md:8]: Just wondering, why here instead of in `Concepts`?

File: docs/configuration/indexes.md
Comment (zanieb) [reply docs/configuration/indexes.md:23]: I think project configuration always takes precedence over configuration outside the project.

File: docs/configuration/indexes.md
Comment (zanieb) [inline docs/configuration/indexes.md:34]: Interesting thing missing here. What if I want to disable having a default index entirely?

File: docs/configuration/indexes.md
Comment (zanieb) [inline docs/configuration/indexes.md:57]: Is this supposed to be

> ... to ensure that _only_ `torch` is installed from ...

File: docs/configuration/indexes.md
Comment (zanieb) [inline docs/configuration/indexes.md:70]: As far as I can tell, we silently ignore if we receive a command-line argument that interacts with a named index right?

Separately, why this limitation? What happened to being able to override a named index on a single invocation?

File: docs/pip/compatibility.md
Comment (zanieb) [inline docs/pip/compatibility.md:151]: Should we refer to "searching across multiple indexes" and deduplicate the content above?

File: crates/distribution-types/src/index.rs
Comment (zanieb) [reply crates/distribution-types/src/index.rs:117]: Sorry I was thinking about `@` — anyway yeah this seems right.

File: docs/configuration/indexes.md
Comment (charliermarsh) [reply docs/configuration/indexes.md:70]: > Separately, why this limitation? What happened to being able to override a named index on a single invocation?

It's an implementation detail... It probably should work, but it's hard because those get merged into settings, and we don't look in the settings when resolving explicit, named indexes (we just look in the current workspace).

File: docs/configuration/indexes.md
Comment (charliermarsh) [reply docs/configuration/indexes.md:23]: Correct.

File: docs/configuration/indexes.md
Comment (charliermarsh) [reply docs/configuration/indexes.md:82]: I think you would need to make it a direct dependency.

File: crates/pypi-types/src/requirement.rs
Comment (charliermarsh) [reply crates/pypi-types/src/requirement.rs:322]: This is correct as a URL. This isn't the user-facing type. But the comment is wrong.

File: crates/uv/tests/lock.rs
Comment (charliermarsh) [reply crates/uv/tests/lock.rs:11897]: The CLI wins!

File: docs/configuration/indexes.md
Comment (charliermarsh) [reply docs/configuration/indexes.md:34]: `--no-index` does that, but otherwise there's no such thing.

File: docs/configuration/index.md
Comment (charliermarsh) [reply docs/configuration/index.md:8]: I'm not sure. To me it was similar to authentication.

File: crates/uv/tests/lock.rs
Comment (charliermarsh) [reply crates/uv/tests/lock.rs:11897]: Added `lock_repeat_named_index_cli` for this.

File: docs/pip/compatibility.md
Comment (charliermarsh) [reply docs/pip/compatibility.md:151]: I could go either way. I don't mind having dedicated content here on how it differs from pip.

File: crates/uv-cli/src/lib.rs
Comment (charliermarsh) [inline crates/uv-cli/src/lib.rs:3756]: @zanieb -- Should we... hide these?

File: crates/uv-cli/src/lib.rs
Comment (charliermarsh) [inline crates/uv-cli/src/lib.rs:3758]: @zanieb -- Should we... hide these?

Comment (zanieb) [reply crates/uv-cli/src/lib.rs:3756]: In a future release, I think.

File: docs/configuration/indexes.md
Comment (charliermarsh) [reply docs/configuration/indexes.md:70]: I'm working on this, but it will be a separate PR.

Comment (charliermarsh) [reply docs/configuration/indexes.md:70]: I'm considering tracking the "source" of every index (CLI, configuration file, etc.), as field on the index type.

File: docs/configuration/indexes.md
Comment (zanieb) [reply docs/configuration/indexes.md:34]: I think it'll be fairly common to want to disable PyPI entirely?

**Review by zanieb — APPROVED** [review]

File: docs/configuration/indexes.md
Comment (charliermarsh) [reply docs/configuration/indexes.md:34]: That’s the same as defining a different default index.

Comment (zanieb) [reply docs/configuration/indexes.md:34]: Yes, but it seems weird to need to do that. I might not want to mark my other index as a default for some reason? I don't fully understand the trade-offs. We can wait and see if it comes up, it doesn't block this pull request.

Comment (charliermarsh) [reply docs/configuration/indexes.md:34]: Yeah, I think what you're getting at here is that `default = true` does two things: (1) replaces PyPI, (2) changes priority.

**Comment by szobov** [comment]
Hey @charliermarsh
Thanks a lot for the work you're doing to improve python's ecosystem.
I'm very much looking forward for this feature since I'm currently blocked by it.

I wanted to ask what are the steps to get it merged and released and if I can somehow assist you?

I tested it locally by installing `uv` from the branch and it worked smoothly :tada:
<details>
<summary>details on the setup</summary>

    ```
    $ uv version
    uv 0.4.18 (f9ee2a9e2 2024-10-03)
    
    $ cat pyproject.toml
    [project]
    name = "inference"
    version = "0.1.0"
    description = "Add your description here"
    readme = "README.md"
    requires-python = ">=3.12"
    dependencies = [
        "numpy>=2.1.2",
        "torch==2.4.0+cu118",
    ]

    [build-system]
    requires = ["hatchling"]
    build-backend = "hatchling.build"

    [tool.uv.sources]
    torch = { index = "pytorch" }

    [[tool.uv.index]]
    name = "pytorch"
    url = "https://download.pytorch.org/whl/cu118"
    explicit = true
    
    $ cat uv.lock
    ....
    [[package]]
    name = "torch"
    version = "2.4.0+cu118"
    source = { registry = "https://download.pytorch.org/whl/cu118" }
    dependencies = [
        { name = "filelock" },
        { name = "fsspec" },
        { name = "jinja2" },
        { name = "networkx" },
        { name = "nvidia-cublas-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-cuda-cupti-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-cuda-nvrtc-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-cuda-runtime-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-cudnn-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-cufft-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-curand-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-cusolver-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-cusparse-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-nccl-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "nvidia-nvtx-cu11", marker = "platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "setuptools" },
        { name = "sympy" },
        { name = "triton", marker = "python_full_version < '3.13' and platform_machine == 'x86_64' and platform_system == 'Linux'" },
        { name = "typing-extensions" },
    ]
    wheels = [
        { url = "https://download.pytorch.org/whl/cu118/torch-2.4.0%2Bcu118-cp312-cp312-linux_x86_64.whl", hash = "sha256:d38bd98e5faf9565f04d2b59a481cf576cdf4078444cdf24868fbf5ad685dc4d" },
        { url = "https://download.pytorch.org/whl/cu118/torch-2.4.0%2Bcu118-cp312-cp312-win_amd64.whl", hash = "sha256:bd16a12a003fe58276d7b7394a3760d576eb8dbfb4bacea025eb52a1eaf5172b" },
    ]
    ```
</details>

Thanks in advance :pray:

**Comment by charliermarsh** [comment]
I’m planning to release it early next week. Sorry for the delay — I’m on vacation today, then at EuroRust later in the week.

**Comment by pythonbyte** [comment]
Hey @charliermarsh 

Amazing work, would be amazing if in this PR we could have the 
`uv add should automatically write --index entries to the pyproject.toml file.`

That would help my team and me not to add a library to write/dump the data.toml file 🙏🏽

**Comment by odie5533** [comment]
> Hey @charliermarsh
> 
> Amazing work, would be amazing if in this PR we could have the `uv add should automatically write --index entries to the pyproject.toml file.`
> 
> That would help my team and me not to add a library to write/dump the data.toml file 🙏🏽

See PRs #7746 and #7747

**Comment by pythonbyte** [comment]
> > Hey @charliermarsh
> > Amazing work, would be amazing if in this PR we could have the `uv add should automatically write --index entries to the pyproject.toml file.`
> > That would help my team and me not to add a library to write/dump the data.toml file 🙏🏽
> 
> See PRs #7746 and #7747

That's awesome, thank you so much, really excited about these features 😄

**Comment by danieleades** [comment]
does this support, or is support planned for, using API keys for authentication against private registries (ie artifactory) rather than using username/password?

**Comment by zanieb** [comment]
You can use an API key in the PASSWORD variable — it works for anything that supports HTTP Basic Authentication.

**Comment by danieleades** [comment]
> You can use an API key in the PASSWORD variable — it works for anything that supports HTTP Basic Authentication.

thanks for the support!

i'm trying this now and seeing issues. For context, i have a project which has one dependency from a private artifactory registry.

with or without credentials exported to the environment i see the following output:

```
uv sync
  × No solution found when resolving dependencies:
  ╰─▶ Because my-private-package was not found in the package registry and your project depends on my-private-package>=2.0.0,<2.1.dev0, we can conclude that your project's requirements are unsatisfiable.

      hint: my-private-package was requested with a pre-release marker (e.g., sb-qag-sphinx>=2.0.0,<2.1.dev0), but pre-releases weren't enabled (try: `--prerelease=allow`)
```

the dependencies are specified as follows:

```toml
dependencies = [
    "my-private-package~=2.0.0",
    # ...
]

[tool.uv.sources]
my-private-package = {index = "mycompany-py-release"}

[[tool.uv.index]]
name = "artifactory-pypi"
url = "https://artifactory.mycompany.com/artifactory/api/pypi/pypi/simple"
default = true

[[tool.uv.index]]
# Requires:
name = "mycompany-py-release"
url = "https://artifactory.mycompany.com/artifactory/api/pypi/my-private-package/simple"
explicit = true
```

with

```sh
export UV_INDEX_MYCOMPANY_PY_RELEASE_USERNAME={username}
export UV_INDEX_MYCOMPANY_PY_RELEASE_PASSWORD={artifactory token}
```

**Comment by zanieb** [comment]
You should open a new issue for this and we can discuss there

**Comment by dariocurr** [comment]
> * Users should be able to specify a proxy URL for a given index, to avoid writing user-specific URLs to a lockfile:
> 
> ```toml
> [[tool.uv.index]]
> name = "test"
> url = "https://private.org/simple"
> proxy = "http://<omitted>/pypi/simple"
> ```

Any news about this feature? This is critical for some users

**Pull request merged**
